src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrQueryInformationDomain2Request.java

```diff
@@ -0,0 +1,91 @@
+import com.rapid7.client.dcerpc.mslsad.objects.DomainInformationClass;
+import com.rapid7.client.dcerpc.mssamr.objects.DomainHandle;
+import com.rapid7.client.dcerpc.mssamr.objects.SAMPRDomainLockoutInfo;
+
+/**
+ * <a href="https://msdn.microsoft.com/en-us/library/cc245773.aspx">SamrQueryInformationDomain2</a>
+ *
+ * <blockquote>
+ * <pre>
+ * The SamrQueryInformationDomain2 method obtains attributes from a domain object.
+ *          long SamrQueryInformationDomain2(
+ *             [in] SAMPR_HANDLE DomainHandle,
+ *             [in] DOMAIN_INFORMATION_CLASS DomainInformationClass,
+ *             [out, switch_is(DomainInformationClass)]
+ *             PSAMPR_DOMAIN_INFO_BUFFER* Buffer
+ *          );
+ * DomainHandle: An RPC context handle, as specified in section 2.2.3.2, representing a domain object.
+ * DomainInformationClass: An enumeration indicating which attributes to return. See section 2.2.4.16 for a listing of possible values.
+ * Buffer: The requested attributes on output. See section 2.2.4.17 for structure details.
+ *
+ * This protocol asks the RPC runtime, via the strict_context_handle attribute, to reject the use of context handles created by a method of a
+ * different RPC interface than this one, as specified in [MS-RPCE] section 3.
+ * Upon receiving this message, the server MUST process the data from the message subject to the following constraints:
+ *
+ *   1.The server MUST return an error if DomainHandle.HandleType is not equal to "Domain".
+ *   2.DomainHandle.GrantedAccess MUST have the required access specified in section 3.1.2.1. Otherwise, the server MUST return STATUS_ACCESS_DENIED.
+ *   3.The following information levels MUST be processed by setting the appropriate output field name to the associated database attribute, as specified
+ *     in section 3.1.5.14.8. Processing is completed by returning 0 on success.
+ * </pre>
+ * </blockquote>
+ */
+public abstract class SamrQueryInformationDomain2Request<T extends Unmarshallable>
+        extends RequestCall<SamrQueryInformationDomainResponse<T>> {
+    public final static short OP_NUM = 46;
+    private final DomainHandle handle;
+    private final DomainInformationClass infoLevel;
+
+    public SamrQueryInformationDomain2Request(final DomainHandle handle, final DomainInformationClass infoLevel) {
+        super(OP_NUM);
+        this.handle = handle;
+        this.infoLevel = infoLevel;
+    }
+
+    abstract T newDomainInformation();
+
+    @Override
+    public SamrQueryInformationDomainResponse<T> getResponseObject() {
+        return new SamrQueryInformationDomainResponse<T>(newDomainInformation(), infoLevel);
+    }
+
+    @Override
+    public void marshal(PacketOutput packetOut) throws IOException {
+        packetOut.write(handle.getBytes());
+        packetOut.writeShort(infoLevel.getInfoLevel());
+    }
+
+    public static class DomainLockoutInformation extends SamrQueryInformationDomain2Request<SAMPRDomainLockoutInfo> {
```

Comment:
Is a static class like this necessary for this?

Reply:
Feels a bit too verbose if we have to call it like below

`        SamrQueryInformationDomain2Request.DomainLockoutInformation request2 = new SamrQueryInformationDomain2Request.DomainLockoutInformation(`

Reply:
@bli-r7 , for example, do you prefer the simplified code as following?
```
  public SamrQueryInformationDomainResponse<SAMPRDomainLockoutInfo> getDomainLockoutInfo(
        final DomainHandle DomainHandle) throws IOException {
       SamrQueryInformationDomain2Request request = new SamrQueryInformationDomain3Request(
            DomainHandle, DomainInformationClass.DOMAIN_LOCKOUT_INFORMATION);
        return transport.call(request);
    }
```
```
public class SamrQueryInformationDomain2Request
        extends RequestCall<SamrQueryInformationDomainResponse<SAMPRDomainLockoutInfo>> {
    public final static short OP_NUM = 46;
    private final DomainHandle handle;
    private final DomainInformationClass infoLevel;

    public SamrQueryInformationDomain3Request(final DomainHandle handle, final DomainInformationClass infoLevel) {
        super(OP_NUM);
        this.handle = handle;
        this.infoLevel = infoLevel;
    }

    @Override
    public void marshal(PacketOutput packetOut) throws IOException {
        packetOut.write(handle.getBytes());
        packetOut.writeShort(infoLevel.getInfoLevel());
    }

    @Override
    public SamrQueryInformationDomainResponse<SAMPRDomainLockoutInfo> getResponseObject() {
        return new SamrQueryInformationDomainResponse<SAMPRDomainLockoutInfo>(new SAMPRDomainLockoutInfo(), infoLevel);
    }

}
```

Reply:
Yes.